Add recursive support for ACL module

astorije opened the pull request:
This PR enables the support for recursive ACLs. Some refactoring was needed to make this work. It is the new repo structure augmented clone of ansible/ansible#8380.

It fixes ansible/ansible#5053, ansible/ansible#5550 and fix ansible/ansible#7276) as well as a few things along the way:
- PEP8 compliance(-ish)
- Fix erroneous entry param `etype` to `type`
- Param checking messages use RFC 2119 wording)
- Add checks for incompatible states

I did not specify the version in the documentation bit. I will add it when this is merged to have consistent docs.
## Advantages of this PR
- Permissions do not need to be normalized anymore using the `--test` checking system
- `--test` does not check using `getfacl`, so the code is less verbose and less likely to introduce a parsing error (since there is actually - almost - no parsing)
- `--test` makes `--recursive` possible at no performance cost, and checking the `changed` status is neither overly complicated or slow
- `default` can now be used on acl queries as well
- [`X` (`capital-x` / _Special execute_)](https://en.wikipedia.org/wiki/Chmod#Symbolic_modes) is now supported in the permissions with no extra code
## Inconvenients of this PR
- `--test` does not seem to be supported by BSD's `setfacl`. However, this can be solved in the `acl_changed` function only without affecting the rest of the file. Suggestions below.
- `recursive=yes` and `state=query` are currently incompatible because that needs further discussion to ensure backward compatibility. This should be done in another issue/PR though. I will be more than happy to work on it, but one step at a time :)
## General notes (and motivations) to help for the review
- `normalize_permissions` did not support `X` (`capital-x`) and its support would cause problem with the former checking system (since `X` itself is not stateless)
- `_run_acl` was making the usage of `--test` impossible because for the last line returned when using `--test` is relevant and not empty. It now makes sure that the last line is empty before stripping it.
- 2 new builder helpers `build_entry` and `build_command` now factorize some recurring code
- `acl_changed` is a function meant to test if a command changes ACLs. The checking can be extended there rather than in the `main` function
- I removed the `get_acls`, `set_acl` and `rm_acl` wrappers for the following reasons:
  - They were used only once each
  - They only contained one call to `build_command` and one call to `run_acl` in the end
  - The command is already built when it is `--test`ed before so I reuse the command rather than rebuilding it
- `_run_acl` is now `run_acl` because it is not a hidden/private function anymore, it is used in the `main` function multiple times
## Suggestions for Linux / BSD support

My PR breaks support for BSD because of its lack of `--test`, but adds support for `--recursive` on Linux. I suggest that when `--test` is not available, `changed` is **always** true when using `recursive=yes`. When not using `recursive=yes`, we can either do the same or replicate the old code for checking state. What I described earlier (notably the `X` bit) will still be a problem. I'll talk with @bcoca on IRC about that.

The documentation should then mention the support differences if there are  any (for example no support for `recursive` on BSD).

astorije committed:
Add recursive support for ACL module (fix #5053, fix #5550, fix #7276)

tristanbes commented:
acl module still not supports recursive option ? :'( too bad. Any move on this PR ?

astorije commented:
Still waiting for feedback from the Ansible guys, sorry :/

issei-m commented:
:+1:

tjanez commented:
This is a must have for my scenario, where I want to recursively set ACLs for a dedicated backup user to `rX` on all directories that are to be backed up.

@astorije, I suggest rebasing the PR to fix the merging conflict and then to send an email requesting attention to ansible-devel mailing list.

abadger commented:
We also have weekly public meetings to work on reviewing and merging the pull requests of people who stop by to discuss them.  If you're available, the meetings are currently every Tuesday from 17:00 UTC to 19:00 UTC.  We send an email to the mailing lists and message to the #ansible and #ansible-devel IRC channels just before the meeting starts with google hangout information (and you can chat on irc during the meeting time too if you can't do a google hangout.)

gregdek commented:
@astorije Thanks for submitting this pull request. Unfortunately, it appears that the underlying code has changed since your submission, and this PR now needs to be rebased. 

If you believe that this PR still applies, please rebase and we will put it into our community review process as soon as possible.

astorije commented:
Oops yes sorry I dropped the ball, @tjanez, @abadger and @gregdek!
I'm on it for the rebase.

astorije commented:
Voilà, @gregdek, rebasing is done :-)
Happy reviewing!

gregdek commented:
Thx @astorije :)

This module is critical to Ansible's functionality, so it's actually owned by the Core team, which is why it's taken a long time to review. Thanks for your patience here.

astorije commented:
No worries! I'll be happy to contribute more when you guys can review and come back with some feedback.

cbrunnkvist commented:
Gah, just ran into this issue. 

Gonna have to work around it with `command:
 setfacl -R -m u:appuser:rwx /mypath` in the meantime.

astorije commented:
@gregdek, I just checked and no rebasing is necessary at the moment. I'll try to stop by maybe once a month to make sure this branch is up-to-date in case I win the review lottery :-)

@cbrunnkvist I'm afraid that's your only choice :-/
If the status of the execution matters to you, I advise you to take a look at the `--test` option.

puiterwijk commented on files/acl.py:
```diff
@@ -151,69 +145,87 @@ def split_entry(entry):
-def get_acls(module,path,follow):
+def build_entry(etype, entity, permissions=None):
+    '''Builds and returns an entry string. Does not include the permissions bit if they are not provided.'''
+    return etype + ':' + entity + (':' + permissions if permissions else '')
```
You cannot use ternary if because that's not in python 2.4

astorije replied:
Good call, I did not know that.

Addressed in f8027a7. Does that seem good/Python enough to you? (Keep in mind that this PR is 1-year old and I haven't done any Python since then :) )

puiterwijk commented on files/acl.py:
```diff
@@ -151,69 +145,87 @@ def split_entry(entry):
+        cmd.append('-x "%s"' % entry)
+    else:  # mode == 'get'
+        cmd = [module.get_bin_path('getfacl', True)]
+        cmd.append('--omit-header')
```
You might want to re-add the warning: # prevents absolute path warnings and removes headers

puiterwijk commented on files/acl.py:
```diff
@@ -151,69 +145,87 @@ def split_entry(entry):
+    cmd = cmd[:]  # lists are mutables so cmd would be overriden without this
+    cmd.insert(1, '--test')
+    lines = run_acl(module, cmd)
+    return not all(line.endswith('*,*') for line in lines)
```
I'm having a hard time parsing this, can you refactor this?

astorije replied:
Hi @puiterwijk, as I was building my defense for this syntax, I noticed that `all()` was Python 2.5+ only, so I replaced it with an ugly `for`-loop in f8027a7. Hope this looks better to you.

About the original syntax though, I do believe (and I checked with a very experienced Python dev colleague of mine) that this was the easiest way to put/read this.
In fact, this makes much more sense if you read it out loud, as this translates in English to:

```
if not all lines end with '*,*' then acls have changed, otherwise acls haven't changed
```

On top of that, I was using a generator so it can cleanly exit prematurely, for efficiency. Not exactly sure what difference that makes here, but assuming one is running the recursive option on a folder of hundred thousands sub-folders and files, on old hardware (after all, we do support Python 2.4 :-) ), it's always nice to terminate early if we can. My `for`-loop keeps this behavior.

puiterwijk commented on files/acl.py:
```diff
@@ -226,79 +238,72 @@ def main():
+            run_acl(module, command, False)
+        msg = "%s is absent" % entry
+
     else:
```
Maybe make this "elif state == 'query' "

astorije replied:
Addressed in f8027a7. Thanks for catching this.

puiterwijk commented on files/acl.py:
```diff
@@ -226,79 +238,72 @@ def main():
     state = module.params.get('state')
     follow = module.params.get('follow')
     default = module.params.get('default')
-
```
Could you add a check if this is linux and do a module.fail_json if not?

astorije replied:
Done in f8027a7.

On one hand, I am very happy we now support recursive ACLs, on the other hand I am very sad we are dropping support for BSD because of this.

Note however that there are some artifacts left: the `-h`option (now on [line 179](https://github.com/ansible/ansible-modules-core/pull/150/files#diff-9c3c3b66bc72c68c9eafd9cefa084c71R179)) is BSD-only while the equivalent must be set with `-L`/`-P` for Linux. See the [Linux man](http://linux.die.net/man/1/setfacl) and the [FreeBSD man](http://www.freebsd.org/cgi/man.cgi?format=html&query=setfacl%281%29) for reference.

I'd be more than happy to give BSD its `acl` module back and to fix that issue described above once this PR makes it upstream. I'm afraid it will be forever delayed if making them part of this PR.

What do you think of my suggestions in the "Suggestions for Linux / BSD support" section of this PR description?

abadger commented:
Cool.  Talked with @puiterwijk as he was reviewing this.  Seems like once his review comments are addressed, this can be merged.

abadger commented on files/acl.py:
```diff
@@ -79,6 +79,14 @@
+    choices: [ 'yes', 'no' ]
+    description:
+      - Recursively sets the specified ACL (added in Ansible @@@). Incompatible with C(state=query).
+
```
If the other review comments get addressed pretty soon, this module should be in 2.0 so you can go ahead and fill that version number in here too.

astorije replied:
Done in f8027a7.

astorije committed:
Ensure Python 2.4 compatibility and Linux-restricted support

- Make build_entry compatible with Python 2.4
- Re-add missing warning/comment that was forgotten while refactoring
- Replace `all()` with a good ol' for-loop Python 2.4 compatibility
- Make a condition check more explicit (when `state` is `query`)
- Make sure this module can only be run with on a Linux distribution
- Add a note about Linux-only support in the documentation
- Set the version in which recursive support was added, 2.0

astorije commented:
Thanks @puiterwijk and @abadger, I'll try to address these ASAP.

astorije commented:
@puiterwijk and @abadger, thanks again for your review. I have now addressed all your comments and, as I figured you'd prefer few commits, I stashed all the fixes into one single commit: f8027a7.

Let me know if you have any further comments and if these are minor, I might be able to address them in a similar timeframe.

bcoca commented on files/acl.py:
```diff
@@ -21,6 +21,8 @@
 description:
      - Sets and retrieves file ACL information.
+notes:
+    - As of Ansible 2.0, this module only supports Linux distributions.
```
is it possible to just make the recursive option ~~BSD~~ Linux only?

astorije replied:
You mean Linux-only, right?

bcoca replied:
yep, that is what i wrote :-)

astorije replied:
Ah ah!

Not really, as everything this refactoring does is based off the `--test` option.

Once this PR makes it upstream, I will definitely work right away on a new PR to fix this. This implies a lot of code though, and this PR already does too much (add/fix capital-`X` support, other fixes, ...).
It would be very sad to see this PR delayed furthermore (initially opened more than a year ago). I promise I'll take care of BSD support right after that. I think I have shown good reactivity (if not always < 24h) every time I was asked, haven't I? :)

astorije replied:
Plus we are talking about a new major version for Ansible (therefore making non-BC changes if following semantic versioning).
When is Ansible 2.0 due? Chances are I'll send a further PR before it gets out.

bcoca replied:
OK, will evaluate this patch as is and work for the 'rebsdization' after it

astorije replied:
Great, thanks a lot! Plus you just find the name of my future PR, thanks :)

astorije committed:
Set version number in recursive option description

astorije commented:
Note that I don't have any test playbooks with me so I did not run the last version of this code...

astorije commented:
Youhouuuuu! Thanks @bcoca!

Okay, to re-BSD-ize this module now, I see a few possible strategies. None of them are perfect, so I'll evaluate them and provide a comparison in a new PR. I remember I did this comparison a year ago, but my memory isn't fresh anymore so I'll work back on this before saying anything wrong.

drzraf commented:
worth noting ? issue  ansible/ansible#12514

jimi-c commented:
@drzraf closed that issue, as it should be opened in this code repo. Thanks!

drzraf commented:
acl: permissions=rX always marked as changed : #2151

astorije commented:
@drzraf, you didn't specify your environment in either ansible/ansible#12514, #2151 or here. I believe this PR fixed this issue along the way (for Linux, support for BSD being still in progress) but without more details it is hard to say.

drzraf commented:
Ansible 2.0.1.0

```
- name: operators permissions                                                                                                                                                             
  acl:                                                                                                                                                                                    
    recursive: yes                                                                                                                                                                        
    name: /srv/www/                                                                                                                                                                       
    etype: group                                                                                                                                                                          
    entity: operator                                                                                                                                                                      
    permissions: rwX                                                                                                                                                                      
    default: yes                                                                                                                                               
    state: present        
```

is always set as "changed"

drzraf commented:
Side note : `default=yes` should be prohibited for regular files

drzraf commented:
Side question: is there a reason `entry` would not support multiple comma-separated entries?
